packages/itwinui-react/src/core/ThemeProvider/ThemeProvider.tsx

```diff
@@ -293,6 +300,63 @@ const useParentThemeAndContext = (rootElement: HTMLElement | null) => {
 
 // ----------------------------------------------------------------------------
 
+/**
+ * Creates a new portal container if necessary, or reuses the parent portal container.
+ *
+ * Updates `portalContainerAtom` with the correct portal container.
+ */
+const PortalContainer = React.memo(
+  ({
+    portalContainerProp,
+    portalContainerFromParent,
+    isInheritingTheme,
+  }: {
+    portalContainerProp: HTMLElement | undefined;
+    portalContainerFromParent: HTMLElement | undefined;
+    isInheritingTheme: boolean;
+  }) => {
+    const [ownerDocument] = useScopedAtom(ownerDocumentAtom);
+    const [portalContainer, setPortalContainer] =
+      useScopedAtom(portalContainerAtom);
+
+    // Synchronize atom with prop changes
+    if (portalContainerProp && portalContainerProp !== portalContainer) {
+      setPortalContainer(portalContainerProp);
+    }
+
+    // bail if not hydrated, because portals don't work on server
+    const isHydrated = useHydration() === 'hydrated';
+    if (!isHydrated) {
+      return null;
+    }
+
+    // Create a new portal container only if necessary:
+    // - not inheriting theme
+    // - no parent portal container to portal into
+    // - parent portal container is in a different window (#2006)
+    if (
+      !portalContainerProp && // bail if portalContainerProp is set, because it takes precedence
+      (!isInheritingTheme ||
+        !portalContainerFromParent ||
+        portalContainerFromParent.ownerDocument !== ownerDocument)
+    ) {
```

Comment:
If we're creating a new `portalContainer` div, I see that we're calling `setPortalContainer` to set the portal container in the store.

But if we're not creating a new `portalContainer` div, should we call `setPortalContainer(portalContainerFromParent)`? Because without it, the third level `ThemeProvider` will create a new `portalContainer` div since it sees that the second level `ThemeProvider` has no `portalContainer` set in its store. I think the correct thing would be for the third level `ThemeProvider` to reuse the div from the first level `ThemeProvider`.

So maybe something like:

```diff
if ( … ) {
  return <div>;
} else {
  setPortalContainer(portalContainerFromParent)
}

…
```

<details>
<summary>Some testing code I tried (App.tsx)</summary>

```jsx
import { Button, ThemeProvider, Tooltip } from '@itwin/itwinui-react';

const App = () => {
  return (
    <>
      <ThemeProvider theme='inherit'>
        <ThemeProvider theme='inherit'>
          <ThemeProvider theme='inherit'>
            <Tooltip content='Tooltip content'>
              <Button>Hello world</Button>
            </Tooltip>
          </ThemeProvider>
        </ThemeProvider>
      </ThemeProvider>
    </>
  );
};

export default App;
```

</details>

Reply:
This should already be handled by the `useScopedAtom` inheritance logic? I will double check

Reply:
Yes, looks like after [`9a66baf`](https://github.com/iTwin/iTwinUI/pull/2011/commits/9a66baf63bcbc0750f3e88c981b63cdfa11f70cb) from https://github.com/iTwin/iTwinUI/pull/2011#discussion_r1583370317, this issue is fixed.